lib/grandpa/grandpa.go

```diff
@@ -17,37 +17,55 @@
 package grandpa
 
 import (
+	"bytes"
+	"sync"
+	"time"
+
 	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/blocktree"
 	"github.com/ChainSafe/gossamer/lib/common"
 	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+
+	log "github.com/ChainSafe/log15"
 )
 
+var interval = time.Second
+
 // Service represents the current state of the grandpa protocol
 type Service struct {
 	// preliminaries
 	blockState BlockState
-	keypair    *ed25519.Keypair //nolint
+	keypair    *ed25519.Keypair
+	mapLock    sync.Mutex
+	chanLock   sync.Mutex
+	stopped    bool
 
 	// current state information
 	state           *State                             // current state
-	subround        subround                           // current sub-round
 	prevotes        map[ed25519.PublicKeyBytes]*Vote   // pre-votes for next state
 	precommits      map[ed25519.PublicKeyBytes]*Vote   // pre-commits for next state
 	pvEquivocations map[ed25519.PublicKeyBytes][]*Vote // equivocatory votes for current pre-vote stage
 	pcEquivocations map[ed25519.PublicKeyBytes][]*Vote // equivocatory votes for current pre-commit stage
 	head            *types.Header                      // most recently finalized block
 
 	// historical information
-	// TODO: do we need maps, or just info from previous round?
+	preVotedBlock      map[uint64]*Vote // map of round number -> pre-voted block
 	bestFinalCandidate map[uint64]*Vote // map of round number -> best final candidate
+
+	// channels for communication with other services
+	in        <-chan *VoteMessage
+	out       chan<- *VoteMessage  // TODO: make this generic
```

Comment:
we need to send `Finalization` messages in the out channel, not just `VoteMessages` 
so this would require adding those messages to the network messages and having them implement the `Message` interface. can open an issue for that

I was thinking about it and I realized I never liked `msgSend` and `msgRec`, I always found them slightly confusing. I think `in` and `out` are shorter and are more clear. if everyone is ok with it it would be nice to change all the service channels to `in` and `out`

Reply:
#933